### PR TITLE
Add Pad op support and fusing fix e2e

### PR DESCRIPTION
**Pull request by ashokkumarkannan1**
**Ticket:**
Fixes https://github.com/tenstorrent/tt-forge-fe/issues/682

 **Description:**
This pull request includes the updated fusing logic where the padding with symmetric pad and mode with constant will be fused to conv2d, added the pad op support , and an updated decomposition for pad modes such as replicate and reflect.

### CODE REVIEW

**Review by nvukobratTT — COMMENTED** [review]
Seems like uneven padding got supported on TTNN. If that is the case, it'll simplify things on our side a bit. 

@ashokkumarkannan1 can you do a check of the state of it? Here si the reference for more details:
- https://github.com/tenstorrent/tt-metal/issues/17656

**Comment by ashokkumarkannan1** [comment]
> Seems like uneven padding got supported on TTNN. If that is the case, it'll simplify things on our side a bit.
> 
> @ashokkumarkannan1 can you do a check of the state of it? Here si the reference for more details:
> 
> * [[Feature Request] Support passing 4 values to padding for Conv2dOp tt-metal#17656](https://github.com/tenstorrent/tt-metal/issues/17656)

Hi @nvukobratTT  checked the ttmetal implmentation of the conv2d op it does support the asymmetric padding cases also,so at initial part there is the issue in the ttir to ttnn conversion part of the conv2dop which result in the legalization issue which is caused due to padlen of 4 in the ttmlir side, currently working on this for the support.

**Comment by nvukobratTT** [comment]
> > Seems like uneven padding got supported on TTNN. If that is the case, it'll simplify things on our side a bit.
> > @ashokkumarkannan1 can you do a check of the state of it? Here si the reference for more details:
> > 
> > * [[Feature Request] Support passing 4 values to padding for Conv2dOp tt-metal#17656](https://github.com/tenstorrent/tt-metal/issues/17656)
> 
> Hi @nvukobratTT checked the ttmetal implmentation of the conv2d op it does support the asymmetric padding cases also,so at initial part there is the issue in the ttir to ttnn conversion part of the conv2dop which result in the legalization issue which is caused due to padlen of 4 in the ttmlir side, currently working on this for the support.

Let's open up issues for MLIR folks to take a look. They're doing some restructuring on conv2d op, so it's better to let them now instead of us to push that change. Can you open up issue and send it back here? :))

**Comment by nvukobratTT** [comment]
> > > > Seems like uneven padding got supported on TTNN. If that is the case, it'll simplify things on our side a bit.
> > > > @ashokkumarkannan1 can you do a check of the state of it? Here si the reference for more details:
> > > > 
> > > > * [[Feature Request] Support passing 4 values to padding for Conv2dOp tt-metal#17656](https://github.com/tenstorrent/tt-metal/issues/17656)
> > > 
> > > 
> > > Hi @nvukobratTT checked the ttmetal implmentation of the conv2d op it does support the asymmetric padding cases also,so at initial part there is the issue in the ttir to ttnn conversion part of the conv2dop which result in the legalization issue which is caused due to padlen of 4 in the ttmlir side, currently working on this for the support.
> > 
> > 
> > Let's open up issues for MLIR folks to take a look. They're doing some restructuring on conv2d op, so it's better to let them now instead of us to push that change. Can you open up issue and send it back here? :))
> 
> Hi @nvukobratTT will raise the ticket at the mlir side this is the workaround that done for your reference https://github.com/tenstorrent/tt-mlir/pull/2816/files

Thanks, let me know once you get that issues in place.

**Comment by ashokkumarkannan1** [comment]
> > > > > Seems like uneven padding got supported on TTNN. If that is the case, it'll simplify things on our side a bit.
> > > > > @ashokkumarkannan1 can you do a check of the state of it? Here si the reference for more details:
> > > > > 
> > > > > * [[Feature Request] Support passing 4 values to padding for Conv2dOp tt-metal#17656](https://github.com/tenstorrent/tt-metal/issues/17656)
> > > > 
> > > > 
> > > > Hi @nvukobratTT checked the ttmetal implmentation of the conv2d op it does support the asymmetric padding cases also,so at initial part there is the issue in the ttir to ttnn conversion part of the conv2dop which result in the legalization issue which is caused due to padlen of 4 in the ttmlir side, currently working on this for the support.
> > > 
> > > 
> > > Let's open up issues for MLIR folks to take a look. They're doing some restructuring on conv2d op, so it's better to let them now instead of us to push that change. Can you open up issue and send it back here? :))
> > 
> > 
> > Hi @nvukobratTT will raise the ticket at the mlir side this is the workaround that done for your reference https://github.com/tenstorrent/tt-mlir/pull/2816/files
> 
> Thanks, let me know once you get that issues in place.

Hi @nvukobratTT   TT_Mlir issue: https://github.com/tenstorrent/tt-mlir/issues/2822

**Comment by nvukobratTT** [comment]
@ashokkumarkannan1 let me know once this PR is ready for review :))

File: forge/forge/op/eval/forge/pad.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+            torch_padding.append(self.padding[i])
+            torch_padding.append(self.padding[i + 1])
+
+        if mode in ["replicate", "reflect"] and (self.pad_len != 4 or self.pad_len != 2):
+            torch_padding = torch_padding[: int(self.pad_len / 2)]
+
+        shape = torch.nn.functional.pad(torch.rand(shape), tuple(torch_padding), mode=mode).shape
+        return tuple(shape), []
+
+    def decompose(self, dc, inputs):
```
Comment (dgolubovicTT) [inline forge/forge/op/eval/forge/pad.py:64]: Do we need decompose now since we have padding supported in MLIR? This decompose logic is copied from pybuda and uses bunch of other ops. I don't think we need this... I we are going to write some new decompose because something is not supported on ttnn, then it should be based on a failing test or something. Do we have such cases?

File: forge/forge/tvm_calls/relay/op/forge_passes.py [hunk 2]
```diff
@@ -186,12 +186,13 @@ def callback(self, pre, post, node_map):
 
         pad_mode = pad.attrs.pad_mode
 
-        if pad_mode == "constant":
-            padding = [top_pad, left_pad, bottom_pad, right_pad]
-
-        if pad_mode == "reflect":
-            act = tvm.relay.op.nn.pad(act, pad_width, pad_mode="reflect")
+        if ((top_pad != bottom_pad or left_pad != right_pad) and (conv_pool.op.name == "nn.max_pool2d")) or (
```
Comment (dgolubovicTT) [inline forge/forge/tvm_calls/relay/op/forge_passes.py:189]: Why this condition for maxpool2d? Can you add comment in the code explaining why all these cases are handled in the same way?

File: forge/forge/tvm_to_python.py [hunk 3]
```diff
@@ -1394,44 +1394,14 @@ def populate_pad_args(graph, nid, compiler_cfg):
...
-    elif len(pad_width_final) == 4:
-        args.append(
-            (
-                "pad",
-                f"({pad_width_final[0]}, {pad_width_final[1]}, {pad_width_final[2]}, {pad_width_final[3]})",
-            )
+    args.append(
+        (
+            "pad",
+            f"({pad_width})",
```
Comment (dgolubovicTT) [inline forge/forge/tvm_to_python.py:1402]: Why we only append pad_width in args? What about other axis?

**Review by dgolubovicTT — CHANGES_REQUESTED** [review]

File: forge/forge/op/eval/forge/pad.py (hunk 1 above)
Comment (ashokkumarkannan1) [reply forge/forge/op/eval/forge/pad.py:64]: Hi @dgolubovicTT  since the MLIR PadOp Only supports padding at the constant mode ,for the other mode we have to follow the decomposition.
Pad Op Mlir: https://docs.tenstorrent.com/tt-mlir/autogen/md/Dialect/TTIROp.html#ttirpad-ttttirpadop

File: forge/forge/tvm_to_python.py (hunk 3 above)
Comment (ashokkumarkannan1) [reply forge/forge/tvm_to_python.py:1402]: Hi @dgolubovicTT  the axis is not needed as the arg for the mlir Pad op.

File: forge/forge/tvm_calls/relay/op/forge_passes.py (hunk 2 above)
Comment (ashokkumarkannan1) [reply forge/forge/tvm_calls/relay/op/forge_passes.py:191]: Added the comments now as suggested @dgolubovicTT

File: forge/csrc/passes/fuse_pad_conv2d.cpp [hunk 4]
```diff
@@ -45,39 +65,46 @@ void fuse_pad_conv2d(graphlib::Graph *graph)
...
         for (auto user : users)
         {
             graphlib::OpNode *user_op = dynamic_cast<graphlib::OpNode *>(user);
-            auto conv_attrs = user_op->op_attrs();
-            TT_ASSERT(conv_attrs.size() == 13);
-            // Conv2d attributes are
-            // [stride[0],stride[1],dilation,groups,padding[0],padding[1],padding[2],padding[3],channel_last]
-            int pad_idx_offset = 4;
-            for (uint32_t i = 0; i < 4; i++)
+            auto conv_attrs = user_op->named_attrs();
```
Comment (dgolubovicTT) [inline forge/csrc/passes/fuse_pad_conv2d.cpp:72]: Nit pick: It is better to leave auto for iterators, lambdas etc. Lets use the right type in these situations.

File: forge/csrc/passes/fuse_pad_conv2d.cpp [hunk 5]
```diff
@@ -45,39 +65,46 @@ void fuse_pad_conv2d(graphlib::Graph *graph)
...
+            TT_ASSERT(conv_attrs.size() == 7);
```
Comment (dgolubovicTT) [inline forge/csrc/passes/fuse_pad_conv2d.cpp:73]: Is there a possibility where size won't be 7?  Or we expect it to be like that all the time in the future?

File: forge/csrc/passes/fuse_pad_conv2d.cpp [hunk 6]
```diff
@@ -45,39 +65,46 @@ void fuse_pad_conv2d(graphlib::Graph *graph)
...
+
+            // Conv2d attributes [stride, dilation, groups, padding]
+            auto &conv_padding_variant = conv_attrs["padding"];
+
+            // Ensure that the padding is a vector
+            std::vector<int> *conv_padding_vec = std::get_if<std::vector<int>>(&conv_padding_variant);
```
Comment (dgolubovicTT) [inline forge/csrc/passes/fuse_pad_conv2d.cpp:79]: What happens if one of the convolutions already had padding? Should we add up paddings alog the axes?

File: forge/forge/op/eval/forge/pad.py (hunk 1 above)
Comment (dgolubovicTT) [inline forge/forge/op/eval/forge/pad.py:64]: Can you please add a detailed comment here how the padding decompose is handled. It is good to have it for anyone who wants to figure out fast what is going on with padding decompose. Or you can add comments in code...

**Review by dgolubovicTT — APPROVED** [review]

File: forge/csrc/passes/fuse_pad_conv2d.cpp (hunk 4 above)
Comment (ashokkumarkannan1) [reply forge/csrc/passes/fuse_pad_conv2d.cpp]: Hi @dgolubovicTT  addressed the above changes.

File: forge/csrc/passes/fuse_pad_conv2d.cpp (hunk 5 above)
Comment (ashokkumarkannan1) [reply forge/csrc/passes/fuse_pad_conv2d.cpp]: Hi @dgolubovicTT  addressed the above changes.

File: forge/csrc/passes/fuse_pad_conv2d.cpp (hunk 6 above)
Comment (ashokkumarkannan1) [reply forge/csrc/passes/fuse_pad_conv2d.cpp:79]: Hi @dgolubovicTT  addressed the above changes.

File: forge/forge/op/eval/forge/pad.py (hunk 1 above)
Comment (ashokkumarkannan1) [reply forge/forge/op/eval/forge/pad.py]: Hi @dgolubovicTT  addressed the above changes.
